File: tools/sdk-ai-bots/azure-sdk-qa-bot-knowledge-sync/src/services/MetadataResolver.ts
import { minimatch } from 'minimatch';
import { Metadata, Override } from './ConfigurationLoader';

/**
 * Service for resolving hierarchical metadata with file overrides
 */
export class MetadataResolver {
    /**
     * Resolve metadata for a file using path defaults and override patterns
     * @param relativePath Relative path from repository root
     * @param pathMetadata Default metadata from DocumentationPath
     * @param overrides Override patterns from DocumentationPath
     * @returns Resolved metadata with scope and plane, or null if no metadata
     */
    static resolveMetadata(
        relativePath: string,
        pathMetadata?: Metadata,
        overrides?: Override[]
    ): Metadata | null {
        // If no metadata at all, return null (will not add metadata fields)
        if (!pathMetadata) {
            return null;
        }

        // Start with path-level defaults
        let metadata: Metadata = { ...pathMetadata };

        // Apply overrides if they exist
        if (overrides && overrides.length > 0) {
            for (const override of overrides) {
                if (this.matchPattern(relativePath, override.pattern)) {
                    // Always use inherit strategy: only update specified fields
                    metadata = {
                        ...metadata,
                        ...override.metadata
                    };
                }
            }
        }

        // Return resolved metadata (scope and plane only)
        return metadata;
    }

    /**
     * Match a file path against a glob pattern
     * @param filePath File path to test
     * @param pattern Glob pattern (e.g., '**\/Azure Data Plane Service\/**', '**\/*paging*.md')
     * @returns True if the pattern matches
     */
    private static matchPattern(filePath: string, pattern: string): boolean {
        // Normalize path separators for Windows compatibility
        const normalizedPath = filePath.replace(/\\/g, '/');
        const normalizedPattern = pattern.replace(/\\/g, '/');
        
        return minimatch(normalizedPath, normalizedPattern, {
            dot: true,           // Match dotfiles
            nocase: true,        // Case-insensitive matching
            matchBase: false     // Don't match basename only
        });
    }

    /**
     * Validate metadata structure
     * @param metadata Metadata to validate
     * @returns True if valid
     */
    static validateMetadata(metadata: Metadata): boolean {
        // Scope is required
        if (!metadata.scope || !['branded', 'unbranded'].includes(metadata.scope)) {
            return false;
        }

        // Plane is optional but must be valid if present
        if (metadata.plane && !['data-plane', 'management-plane'].includes(metadata.plane)) {
            return false;
        }

        // Plane only makes sense for branded content
        if (metadata.scope === 'unbranded' && metadata.plane) {
            console.warn('Warning: plane is set for unbranded content and will be ignored');
        }

        return true;
    }
}

File: tools/sdk-ai-bots/azure-sdk-qa-bot-knowledge-graph-sync/src/services/metadata_resolver.py
"""Metadata resolver — resolves scope and service_type for documents.

Uses glob pattern matching to apply file-level overrides on top of
path-level default metadata.
"""

from __future__ import annotations

import logging

from wcmatch import fnmatch

from src.services.configuration_loader import Metadata, Override

logger = logging.getLogger(__name__)


class MetadataResolver:
    """Resolves hierarchical metadata with file overrides."""

    @staticmethod
    def resolve_metadata(
        relative_path: str,
        path_metadata: Metadata | None,
        overrides: list[Override] | None = None,
    ) -> Metadata | None:
        """Resolve metadata for a file using path defaults and overrides.

        Args:
            relative_path: Relative path from source root
            path_metadata: Default metadata from DocumentationPath
            overrides: Per-file override rules

        Returns:
            Resolved Metadata, or None if no metadata applies
        """
        if path_metadata is None:
            return None

        # Start with path-level defaults
        resolved = Metadata(
            scope=path_metadata.scope,
            service_type=path_metadata.service_type,
        )

        # Apply overrides (last match wins for each field)
        if overrides:
            for override in overrides:
                if MetadataResolver._match_pattern(relative_path, override.pattern):
                    if override.metadata.scope:
                        resolved.scope = override.metadata.scope
                    if override.metadata.service_type:
                        resolved.service_type = override.metadata.service_type

        return resolved

    @staticmethod
    def validate_metadata(metadata: Metadata) -> bool:
        """Validate metadata structure."""
        if metadata.scope not in ("branded", "unbranded"):
            return False
        if metadata.service_type and metadata.service_type not in (
            "data-plane",
            "management-plane",
        ):
            return False
        if metadata.scope == "unbranded" and metadata.service_type:
            logger.warning("service_type is set for unbranded content and will be ignored")
        return True

    @staticmethod
    def _match_pattern(file_path: str, pattern: str) -> bool:
        """Match a file path against a glob pattern."""
        # Normalize path separators
        normalized_path = file_path.replace("\\", "/")
        normalized_pattern = pattern.replace("\\", "/")

        return fnmatch.fnmatch(
            normalized_path,
            normalized_pattern,
            flags=fnmatch.DOTMATCH | fnmatch.IGNORECASE | fnmatch.GLOBSTAR,
        )
